### PR TITLE
Add option to rotate image (#5)

### PR DESCRIPTION
**Pull request by Arcadie**
Use this option for ascending (South -> North) satellite passes.
It will rotate the image and make sure channels A and B are not swapped.

Use '--rotate-image' argument on the command line or
'Rotate output image (South -> North passes)' in 'Advanced settings' GUI.

Note: my Rust is really _rusty_, so please double check the rotation part. 
I reversed the signal array, shifted 1/2 row length from the end to preserve location of channels A and B, but since the telemetry is embedded in the signal itself, the notches are on the wrong side.

Not sure if this matters since the image is for display only, though.
Tested on `argentina.wav` and a couple of my captures, both with sync and without.
Seems to be working as expected.

Thanks for the awesome tool!

### CODE REVIEW

**Comment by martinber** [comment]
Thank you very much! The GUI part looks good

You got wrong the part where you move the A channel to the left:

```rust
let mut rotated_image = signal.clone();
rotated_image.reverse();

// Now channels A and B have swapped places, need to fix this
// by taking the last PX_PER_ROW / 2 bytes and making sure they end up at the beginning:
let (first, last) = rotated_image.split_at(signal.len() - (PX_PER_ROW / 2) as usize);
let fixed_image: Vec<u8> = last.iter().chain(first.iter()).map(|&x| x).collect();
writer.write_image_data(&fixed_image[..])?;
```

At this point, the `signal`,  `rotated_image` and `fixed_image` are one-dimensional `Vec`s of pixel values. When written to the PNG file with `writer.write_image_data();` this is transformed into a two-dimensional image by writing line by line.

The `signal`,  `rotated_image` and `fixed_image` have alternating chains of A and B channel pixels: 1040 A pixels, 1040 B pixels, 1040 B pixels, 1040 A pixels... (1040 = PX_PER_ROW/2)

The channel on the left of the final image will be the first one on the `fixed_image` chain:

path1953-1

What you did works: you are taking the last 1040 values of the `rotated_image` (corresponding to A channel) and appending them to the start. So now the fixed image has B values at the start. The thing is that it does not make sense to do this, is like taking the bottom of the image and putting it at the start:

rect3824-2

Instead of moving values around, I would just delete the first 1040 pixels (B channel) from `rotated_image`. And also the last 1040 pixels (A channel) so the length of `fixed_image` is a multiple of 2080 (PX_PER_ROW). You can do this by using slices.

As you can see it's a minor thing. I'm explaining it instead of doing it myself just in case you want to have fun with it. Otherwise just leave it, In that case I will do it myself next week.

File: src/noaa_apt.rs [hunk 1]
```diff
@@ -380,11 +380,26 @@ pub fn decode(
     encoder.set_depth(png::BitDepth::Eight);
     let mut writer = encoder.write_header()?;
 
-    writer.write_image_data(&signal[..])?;
+    if settings.rotate_image {
+        context.status(0.98, "Rotating output image".to_string());
+        debug!("Rotating image");
+
+        let mut rotated_image = signal.clone();
+        rotated_image.reverse();
```
Comment (oherrala) [inline src/noaa_apt.rs]: You can probably eliminate the clone here by using iterator:

```rust
let rotated_image = signal.iter().rev();
```

and then on line 392 you continue with the same iterator:

```rust
let first_part = rotated_image.take(signal.len() - (PX_PER_ROW / 2) as usize));
let fixed_image: Signal = rotated_image.chain(first_part).copied().collect();
```

didn't test this, but probably should work something like that.

**Comment by Arcadie** [comment]
@martinber, thank you for the amazing explanation and images, it made it really easy to understand how the data is structured and what is going on.

I've thought about this, and one thing that bothers me when shifting or slicing the signal is that channels A and B are no longer _in sync_ when looking at each line:

```
signal:
A1B1               
A2B2
A3B3

reversed:
B3A3
B2A2
B1A1

fixed:
A3B2    <- channel A from line 3 shows up next to channel B from line 2
A2B1    <- channel A from line 2 shows up next to channel B from line 1
```

I don't know how much of an issue this might turn out to be, but it would be nice to avoid this if possible.

I've found a way to rotate the image and swap channels A and B without the need to shift / drop samples, at the expense of partitioning the channels and then interleaving them again in the correct order.

Please let me know if this is an acceptable solution for you. Thanks!

File: src/noaa_apt.rs (hunk 1 above)
Comment (Arcadie) [reply src/noaa_apt.rs]: @oherrala, thanks! I did use `.iter().rev()` in this new revision thanks to your suggestion.

**Comment by martinber** [comment]
I like the idea of interleaving the samples, at first I did not realize that my solution would create an offset between both channels. It is small but it could create problems if someone decides to do further processing.

I'm going to merge it, and if you don't mind I'm going to try to get to a better solution. You code gave me some ideas.

Thank you very much to both of you! I'm adding @Arcadie as Arcadie Z. to the contributor list, if you want to be added as a different name just ask

**Comment by Arcadie** [comment]
> as Arcadie Z. to the contributor list

Not required, but much appreciated, thanks! 😄 
Thanks again for the awesome decoder, I'm using it almost every day.

**Pull request merged**